server/data/data-jmap/src/main/java/org/apache/james/jmap/api/change/JmapChangeFactory.java

```diff
@@ -0,0 +1,292 @@
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.api.change;
+
+import java.time.Clock;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.mail.Flags;
+
+import org.apache.james.core.Username;
+import org.apache.james.jmap.api.model.AccountId;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.events.Event;
+import org.apache.james.mailbox.events.MailboxListener;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxACL;
+import org.apache.james.mailbox.model.MailboxId;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
+public class JmapChangeFactory {
+    private static Stream<String> getSharees(MailboxId mailboxId, Username username, MailboxManager mailboxManager) {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(username);
+        try {
+            MailboxACL mailboxACL = mailboxManager.listRights(mailboxId, mailboxSession);
+            return mailboxACL.getEntries().keySet()
+                .stream()
+                .filter(rfc4314Rights -> !rfc4314Rights.isNegative())
+                .filter(rfc4314Rights -> rfc4314Rights.getNameType().equals(MailboxACL.NameType.user))
+                .map(MailboxACL.EntryKey::getName);
+        } catch (MailboxException e) {
+            return Stream.of();
+        }
+    }
+
+    private final Clock clock;
+    private final MailboxManager mailboxManager;
+    private final State.Factory stateFactory;
+
+    @Inject
+    public JmapChangeFactory(Clock clock, MailboxManager mailboxManager, State.Factory stateFactory) {
+        this.clock = clock;
+        this.mailboxManager = mailboxManager;
+        this.stateFactory = stateFactory;
+    }
+
+    public List<JmapChange> fromEvent(Event event) {
```

Comment:
If you want to avoid a double read on ACLs do it before, and avoid mixing things IMO.

```
public List<EmailChange> fromEvent(Event event, MailboxACL acl); // In the factory for emailChange factory

public List<MailboxChange> fromEvent(Event event, MailboxACL acl);
```

If it is not enough you can pass the list of sharee instead?

Reply:
200+ loc methods are always a bad idea, we should break things in smaller methods `fromAdded`, `fromExpunged`...